oak/server/rust/oak_runtime/src/runtime.rs

```diff
@@ -0,0 +1,136 @@
+    terminating: AtomicBool,
+    node_threads: Mutex<Vec<std::thread::JoinHandle<()>>>,
+}
+
+impl Runtime {
+    /// Configure and run the protobuf specified `ApplicationConfiguration`. After creating a
+    /// `Runtime` calling `stop` will send termination signals to nodes and wait for them to
+    /// terminate.
+    pub fn configure_and_run(
+        app_config: proto::manager::ApplicationConfiguration,
+    ) -> Result<(RuntimeRef, ChannelWriter), OakStatus> {
+        let mut runtime = Runtime {
+            configurations: HashMap::new(),
+            terminating: AtomicBool::new(false),
+            node_threads: Mutex::new(Vec::new()),
+        };
+
+        for config in app_config.get_node_configs() {
+            runtime.configurations.insert(
+                config.name.clone(),
+                match &config.config_type {
+                    None => {
+                        error!("Node config {} with no type", config.name);
+                        return Err(OakStatus::ERR_INVALID_ARGS);
+                    }
+                    Some(NodeConfiguration_oneof_config_type::log_config(_)) => {
+                        NodeConfiguration::LogNode
+                    }
+                    Some(NodeConfiguration_oneof_config_type::wasm_config(
+                        proto::manager::WebAssemblyConfiguration { module_bytes, .. },
+                    )) => load_wasm(&module_bytes),
+                    Some(_) => {
+                        error!("Pseudo-node not implemented!");
+                        return Err(OakStatus::ERR_INVALID_ARGS);
+                    }
+                },
+            );
+        }
+
+        let runtime = RuntimeRef(Arc::new(runtime));
+        let (chan_writer, chan_reader) = channel::new();
+
+        runtime.node_create(
+            &app_config.initial_node_config_name,
+            &app_config.initial_entrypoint_name,
+            chan_reader,
+        )?;
+
+        Ok((runtime, chan_writer))
+    }
+
+    /// Thread safe method for determining if the `Runtime` is terminating.
+    pub fn is_terminating(&self) -> bool {
+        self.terminating.load(SeqCst)
+    }
+
+    /// Thread safe method for signaling termination to a `Runtime` and waiting for it's node
+    /// threads to terminate.
+    pub fn stop(&self) {
+        let mut node_threads = self.node_threads.lock().unwrap();
+
+        self.terminating.store(true, SeqCst);
+
+        let mut handles = vec![];
+        std::mem::swap(&mut handles, &mut node_threads);
+
+        for handle in handles {
+            handle.join().expect("Failed to join handle");
+        }
+    }
+}
+
+/// A reference to a `Runtime`
+#[derive(Clone)]
+pub struct RuntimeRef(Arc<Runtime>);
+
+impl RuntimeRef {
+    /// Thread safe method that attempts to create a node within the `Runtime` corresponding to a
+    /// given module name and entrypoint. The `reader: ChannelReader` is passed to the newly
+    /// created node.
+    pub fn node_create(
```

Comment:
What is the advantage of this vs declaring it directly on `Runtime`? Perhaps add a comment for future reference?

Reply:
`conf.new_instance` took an `Arc<Runtime>` previously, its been updated to `RuntimeRef` (which is just a `Arc<Runtime>` wrapper). `conf.new_instance` needs an `Arc` so the node thread can reference the `Runtime`. The method can be moved into the `impl Runtime {` stanza, but it would be a standalone method (not `&self`), e.g.

~~~rust
    pub fn node_create(
        runtime: Arc<Runtime>,
        module_name: &str,
        entrypoint: &str,
        reader: ChannelReader,
    ) -> Result<(), OakStatus> {
        ...
~~~

Reply:
Added comment